// @pulsoid/socket/dist/index.es.js
const d = (i, e) => {
  let t = null;
  const n = ((...o) => {
    t !== null && clearTimeout(t), t = setTimeout(() => {
      i(...o), t = null;
    }, e);
  });
  return n.cancel = () => {
    t !== null && (clearTimeout(t), t = null);
  }, n;
}, r = class r {
  constructor(e, t = {}) {
    this.token = e, this.websocket = null, this.reconnectTryCount = 0, this.reconnectTimeout = null, this.hasOpened = !1, this.lastCloseEvent = null, this.onOpenEventHandler = (n) => {
      this.hasOpened = !0, this.emitEvent("open", n), this.reconnectTryCount > 0 && this.resetReconnectData();
    }, this.onCloseEventHandler = (n) => {
      this.lastCloseEvent = n, this.reconnectTryCount > 0 || this.onFirstCloseCleanup(), this.shouldReconnect && this.reconnectTryCount < this.reconnectOptions.reconnectAttempts ? this.reconnect() : this.emitCloseAndCleanup(), this.shouldReconnect = this.reconnectOptions.enable;
    }, this.onErrorEventHandler = (n) => {
      this.emitEvent("error", n);
    }, this.onReconnectEventHandler = () => {
      this.emitEvent("reconnect", {
        attempt: this.reconnectTryCount
      });
    }, this.boundOnMessageHandler = (n) => {
      this.onMessageHandler(n);
    }, this.isConnected = () => this.websocket?.readyState === WebSocket.OPEN, this.reconnectOptions = {
      ...r.reconnectDefaultOptions,
      ...t
    }, this.shouldReconnect = this.reconnectOptions.enable, this.eventTypeToEventHandlersMap = {};
    for (const n of this.getEventTypes())
      this.eventTypeToEventHandlersMap[n] = [];
  }
  onFirstCloseCleanup() {
  }
  onDisconnect() {
  }
  assignEventHandlers() {
    this.websocket && (this.websocket.addEventListener("open", this.onOpenEventHandler), this.websocket.addEventListener(
      "close",
      this.onCloseEventHandler
    ), this.websocket.addEventListener(
      "message",
      this.boundOnMessageHandler
    ), this.websocket.addEventListener("error", this.onErrorEventHandler));
  }
  clearEventHandlers() {
    this.websocket && (this.websocket.removeEventListener("open", this.onOpenEventHandler), this.websocket.removeEventListener(
      "close",
      this.onCloseEventHandler
    ), this.websocket.removeEventListener(
      "message",
      this.boundOnMessageHandler
    ), this.websocket.removeEventListener("error", this.onErrorEventHandler));
  }
  resetReconnectData() {
    this.reconnectTryCount = 0, this.reconnectTimeout !== null && (clearTimeout(this.reconnectTimeout), this.reconnectTimeout = null);
  }
  reconnect() {
    this.reconnectTimeout === null && (this.reconnectTimeout = setTimeout(() => {
      this.reconnectTimeout = null, this.reconnectTryCount++, this.onReconnectEventHandler(), this.validateToken().then(() => {
        try {
          this.openSocketConnection();
        } catch {
          this.reconnectTryCount < this.reconnectOptions.reconnectAttempts ? this.reconnect() : this.emitCloseAndCleanup();
        }
      }).catch((e) => {
        this.emitTokenError(e), r.isRetriableTokenError(e) && this.reconnectTryCount < this.reconnectOptions.reconnectAttempts ? this.reconnect() : this.emitCloseAndCleanup();
      });
    }, this.getReconnectInterval()));
  }
  getReconnectInterval() {
    const { reconnectMinInterval: e, reconnectMaxInterval: t } = this.reconnectOptions, n = e * Math.pow(2, this.reconnectTryCount);
    return Math.min(n, t);
  }
  static isRetriableTokenError(e) {
    return e.type === "network_error" || e.type === "payment_required" || e.type === "unknown";
  }
  static getTokenErrorType(e) {
    switch (e) {
      case 401:
        return "unauthorized";
      case 402:
        return "payment_required";
      case 403:
        return "forbidden";
      default:
        return "unknown";
    }
  }
  async validateToken() {
    let e;
    try {
      e = await fetch(
        "https://dev.pulsoid.net/api/v1/token/validate",
        {
          headers: { Authorization: `Bearer ${this.token}` }
        }
      );
    } catch {
      throw {
        type: "network_error",
        code: 0,
        message: "Network request failed. Check your internet connection."
      };
    }
    const t = await e.json();
    if (!e.ok)
      throw {
        type: r.getTokenErrorType(e.status),
        code: t.error_code,
        message: t.error_message
      };
    if (!(t.scopes ?? []).includes(this.requiredScope))
      throw {
        type: "insufficient_scope",
        code: 7008,
        message: `insufficient_scope: required scope "${this.requiredScope}" is missing`
      };
  }
  emitCloseAndCleanup() {
    this.hasOpened && (this.emitEvent("close", this.lastCloseEvent), this.hasOpened = !1), this.clearEventHandlers();
  }
  emitTokenError(e) {
    this.emitEvent("token-error", e);
  }
  openSocketConnection() {
    this.clearEventHandlers(), this.websocket = new WebSocket(this.url), this.assignEventHandlers();
  }
  emitEvent(e, ...t) {
    const n = this.eventTypeToEventHandlersMap[e];
    n && n.forEach((o) => o(...t));
  }
  on(e, t) {
    this.eventTypeToEventHandlersMap[e].push(t);
  }
  off(e, t) {
    t ? this.eventTypeToEventHandlersMap[e] = this.eventTypeToEventHandlersMap[e].filter(
      (n) => n !== t
    ) : this.eventTypeToEventHandlersMap[e] = [];
  }
  async connect() {
    this.websocket?.readyState === WebSocket.OPEN || this.websocket?.readyState === WebSocket.CONNECTING || (await this.validateToken(), this.resetReconnectData(), this.openSocketConnection());
  }
  disconnect() {
    this.onDisconnect(), this.resetReconnectData(), this.shouldReconnect = !1, this.websocket?.readyState === WebSocket.OPEN || this.websocket?.readyState === WebSocket.CONNECTING ? this.websocket.close() : this.emitCloseAndCleanup();
  }
};
r.reconnectDefaultOptions = {
  enable: !0,
  reconnectMinInterval: 2e3,
  reconnectMaxInterval: 1e4,
  reconnectAttempts: 100
};
let a = r;
const u = [
  "heart_rate",
  "room_member_updated",
  "room_member_removed",
  "room_updated"
];
class l extends a {
  constructor(e, t, n, o = {}) {
    super(e, o.reconnect), this.roomId = t, this.kinds = n;
  }
  get requiredScope() {
    return "data:room:read";
  }
  get url() {
    const e = this.kinds.join(",");
    return `wss://dev.pulsoid.net/api/v2/data/rooms/${this.roomId}/real_time?access_token=${this.token}&kinds=${e}`;
  }
  getEventTypes() {
    return [
      "open",
      "close",
      "error",
      "reconnect",
      "token-error",
      "heart-rate",
      "room-member-updated",
      "room-member-removed",
      "room-updated"
    ];
  }
  onMessageHandler(e) {
    try {
      const t = JSON.parse(e.data), { kind: n, timestamp: o } = t;
      switch (n) {
        case "heart_rate": {
          const { profile_id: s, bpm: c } = t.heart_rate;
          this.emitEvent("heart-rate", { profileId: s, bpm: c, timestamp: o });
          break;
        }
        case "room_member_updated": {
          const { profile_id: s, config: c } = t.room_member_updated;
          this.emitEvent("room-member-updated", {
            profileId: s,
            config: c,
            timestamp: o
          });
          break;
        }
        case "room_member_removed": {
          const { profile_id: s } = t.room_member_removed;
          this.emitEvent("room-member-removed", {
            profileId: s,
            timestamp: o
          });
          break;
        }
        case "room_updated": {
          const { room_id: s, config: c } = t.room_updated;
          this.emitEvent("room-updated", {
            roomId: s,
            config: c,
            timestamp: o
          });
          break;
        }
      }
    } catch (t) {
      console.warn("[PulsoidRoomSocket] Malformed message:", e.data, t);
    }
  }
  static create(e, t, n) {
    const o = n?.kinds ?? u;
    return new l(e, t, o, n);
  }
  on(e, t) {
    super.on(e, t);
  }
  off(e, t) {
    super.off(e, t);
  }
}
const m = (i) => {
  try {
    const e = JSON.parse(i), t = e?.data?.heart_rate, n = e?.measured_at;
    return typeof t != "number" || typeof n != "number" ? (console.warn("[PulsoidSocket] Missing or invalid fields:", i), null) : { heartRate: t, measuredAt: n };
  } catch (e) {
    return console.warn("[PulsoidSocket] Malformed message:", i, e), null;
  }
};
class h extends a {
  constructor(e, t = {}) {
    super(e, t.reconnect), this.online = !1, this.debounceOfflineEvent = d(
      () => this.onOfflineEventHandler(),
      3e4
    ), this.onOfflineEventHandler = () => {
      this.online && (this.emitEvent("offline"), this.online = !1);
    }, this.onOnlineEventHandler = () => {
      this.online || (this.emitEvent("online"), this.online = !0), this.debounceOfflineEvent();
    }, this.isOnline = () => this.online;
  }
  get requiredScope() {
    return "data:heart_rate:read";
  }
  get url() {
    return `wss://dev.pulsoid.net/api/v1/data/real_time?access_token=${this.token}`;
  }
  getEventTypes() {
    return [
      "open",
      "heart-rate",
      "error",
      "close",
      "online",
      "offline",
      "reconnect",
      "token-error"
    ];
  }
  onMessageHandler(e) {
    const t = m(e?.data);
    this.onOnlineEventHandler(), t && this.emitEvent("heart-rate", t);
  }
  onFirstCloseCleanup() {
    this.online && (this.onOfflineEventHandler(), this.debounceOfflineEvent.cancel());
  }
  onDisconnect() {
    this.debounceOfflineEvent.cancel();
  }
  static create(e, t) {
    return new h(e, t);
  }
  static createRoom(e, t, n) {
    return l.create(e, t, n);
  }
  on(e, t) {
    super.on(e, t);
  }
  off(e, t) {
    super.off(e, t);
  }
}
export {
  l as PulsoidRoomSocket,
  h as PulsoidSocket,
  h as default
};
